### ISSUE TITLE
RL question

### ISSUE THREAD

**Issue by markict123** [CatIIIIIIII/VeriPrefer#3]
Hi,
I tried to use the data with tb in this article to fine-tune the SFT model. GRPO is employed to perform reinforcement learning. However, it seems that the reward is negative.( if the generated code pass the tb, reward is 1. otherwise 0.)

<img width="1809" height="725" alt="Image" src="https://github.com/user-attachments/assets/044663a3-fabc-4565-bbc3-248996e5ac20" />
And the model fine-tuned with GRPO doesn't show better performance comparing with the SFT model.

I've noticed that PPO brings negative effect according to TABLE VII in this article. Could you share some insights? Have you met the similar situation of negative reward?

**Comment by CatIIIIIIII** [comment]
Hi,

I also failed to perform GRPO on the SFT model before. The key is that the current SFT model has no ability to reason, like the reasoning trace output "Oh, I need to do this, so let's ..". Reinforcement learning with veriable rewards (rlvr, like GRPO, DAPO, ...) only "incentive" the reasoning ability not create. You should perform GRPO on some SFT model that already trained on chain of thought (CoT) data. https://huggingface.co/zhuyaoyu/CodeV-R1-Distill-Qwen-7B maybe a good choice.

Also, please let me know if you tried on the CodeV-R1 model. I am also interested in this.

**Comment by markict123** [comment]
> Hi,
> 
> I also failed to perform GRPO on the SFT model before. The key is that the current SFT model has no ability to reason, like the reasoning trace output "Oh, I need to do this, so let's ..". Reinforcement learning with veriable rewards (rlvr, like GRPO, DAPO, ...) only "incentive" the reasoning ability not create. You should perform GRPO on some SFT model that already trained on chain of thought (CoT) data. https://huggingface.co/zhuyaoyu/CodeV-R1-Distill-Qwen-7B maybe a good choice.
> 
> Also, please let me know if you tried on the CodeV-R1 model. I am also interested in this.

I see. But I tried to perform GRPO with tb from VerilogEval before using DeepSeekCoder 7B as the base model. And the result is effective although the base model doesn't have reasoning ability. It can directly generate the correct output without CoT trace.
Maybe the data from your article is more difficult, so the model has to use reasoning ability?

**Comment by CatIIIIIIII** [comment]
I think it has no relationship with data difficulty. This is because during pretrain, Qwen series models use many reasoning data, so it has some reasoning ability. If you read RLVR related paper, you will find out almost all of them use Qwen2.5 models.

But after fine tune on Verilog dataset, the original reasoning ability is forgotten, which is know as "catastrophic forgetting" in deep learning.

**Comment by markict123** [comment]
> I think it has no relationship with data difficulty. This is because during pretrain, Qwen series models use many reasoning data, so it has some reasoning ability. If you read RLVR related paper, you will find out almost all of them use Qwen2.5 models.
> 
> But after fine tune on Verilog dataset, the original reasoning ability is forgotten, which is know as "catastrophic forgetting" in deep learning.

Why it is considered essential for a model to have reasoning capabilities? If a model simply generates code directly based on instructions without an explicit chain of thought, wouldn’t reinforcement learning still be applicable? 

And from a theoretical standpoint, DPO also has an implicit reward function and, according to the derivations in the DPO paper, is theoretically equivalent to PPO. Why is there such a notable difference between the two approaches in practice?

**Comment by CatIIIIIIII** [comment]
You can view reasoning trace as a kind of exploration. LLM will 'search' the right answer with reasoning. I think RLVR will not work for model without reasoning capabilities, check here https://arxiv.org/abs/2504.13837.

Yes, they are theoretically equivalent. But in practice, give data an objective, universial score is much harder and less precise compared with compare two answers. check https://arxiv.org/pdf/2503.01067

**Comment by markict123** [comment]
> You can view reasoning trace as a kind of exploration. LLM will 'search' the right answer with reasoning. I think RLVR will not work for model without reasoning capabilities, check here https://arxiv.org/abs/2504.13837.
> 
> Yes, they are theoretically equivalent. But in practice, give data an objective, universial score is much harder and less precise compared with compare two answers. check https://arxiv.org/pdf/2503.01067

Thank you for your patient responses and highly insightful suggestions.
